src/ucs/datastruct/bitmap.h

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCS_BITMAP_H_
+#define UCS_BITMAP_H_
+
+#include <ucs/sys/compiler_def.h>
+#include <stdint.h>
+
+BEGIN_C_DECLS
+
+
+#define _UCS_BITMAP_BITS_IN_WORD (64)
+
+
+#define _UCS_BITMAP_WORD_INDEX(_bit_index) ((_bit_index) / _UCS_BITMAP_BITS_IN_WORD)
+
+
+#define _UCS_BITMAP_BIT_IN_WORD_INDEX(_bit_index) ((_bit_index) % _UCS_BITMAP_BITS_IN_WORD)
+
+/*
+ * Represents an n-bit bitmap, where n is a multiple of 64, by using an array
+ * of 64-bit unsigned long integers.
+ * 
+ * @param _length Bits number in the bitmap
+ */
+#define _UCS_BITMAP_DECLARE_TYPE(_length) \
+    typedef struct { \
+        uint64_t bits[_UCS_BITMAP_WORD_INDEX(_length)]; \
+    } ucs_bitmap_t(_length);
+
+
+/**
+ * Expands to bitmap type definition
+ *
+ * @param _length Bits number (as passed to UCS_BITMAP_DECLARE_TYPE)
+ *
+ * Example:
+ *
+ * @code{.c}
+ * ucs_bitmap_t(64) my_bitmap;
+ * @endcode
+ */
+#define ucs_bitmap_t(_length) \
+    _UCS_BITMAP_IDENTIFIER(_length, _t)
+
+
+#define _UCS_BITMAP_IDENTIFIER(_name, _suffix) \
+    UCS_PP_TOKENPASTE(ucs_bitmap_, UCS_PP_TOKENPASTE(_name, _suffix))
+
+
+_UCS_BITMAP_DECLARE_TYPE(64)
+_UCS_BITMAP_DECLARE_TYPE(128)
+_UCS_BITMAP_DECLARE_TYPE(256)
+
+
+/**
+ * Get bit value
+ * 
+ * @param _bitmap Bitmap to get from
+ * @param _index Bit index
+ * 
+ * @return Bit value (0 or 1)
+*/
+#define UCS_BITMAP_GET(_bitmap, _index) \
+    (!!((_bitmap)->bits[_UCS_BITMAP_WORD_INDEX(_index)] & \
+    UCS_BIT(_UCS_BITMAP_BIT_IN_WORD_INDEX(_index))))
+
+
+/**
+ * Set a bit
+ * 
+ * @param _bitmap Bitmap to set in
+ * @param _index Bit index
+ */
+ #define UCS_BITMAP_SET(_bitmap, _index) \
+    ((_bitmap)->bits)[_UCS_BITMAP_WORD_INDEX(_index)] |= \
+    (UCS_BIT(_UCS_BITMAP_BIT_IN_WORD_INDEX(_index)));
+
+
+/**
+ * Unset a bit
+ * 
+ * @param _bitmap Bitmap to unset in
+ * @param _index Bit index
+ */
+ #define UCS_BITMAP_UNSET(_bitmap, _index) \
+     ((_bitmap)->bits)[_UCS_BITMAP_WORD_INDEX(_index)] &= \
+     ~((UCS_BIT(_UCS_BITMAP_BIT_IN_WORD_INDEX(_index))));
+
+
+/**
+ * Clear a bitmap by setting all bits to zero
+ * 
+ * @param _bitmap Bitmap to clear
+ * 
+ * @return Value of the bit (0 or 1)
+ */
+#define UCS_BITMAP_CLEAR(_bitmap) \
+    (memset((_bitmap)->bits, 0, sizeof((_bitmap)->bits)))
+
+
+/**
+ * Iterate over all set bits of a given bitmap
+ * 
+ * @param _bitmap Bitmap to iterate on
+ * @param _index Bit index (global offset - relative to the whole bitmap)
+ * @param _word Index of the 64-bit word in which the bit is set
+ * @param _index_in_word Index of the bit in its 64-bit word
+ */
+#define UCS_BITMAP_FOR_EACH_BIT(_bitmap, _index, _word, _index_in_word) \
```

Comment:
can we avoid requiring the user to define _index_in_word variable, and use just one index?

Reply:
like this:
```
#define _UCS_BITMAP_WORD(_bitmap, _index) ((_bitmap)->bits[_UCS_BITMAP_WORD_INDEX(_index)])
#define _UCS_BITMAP_WORD_INDEX0(_bit_index) ((_bit_index) & ~(_UCS_BITMAP_BITS_IN_WORD - 1))

#define UCS_BITMAP_FOR_EACH_BIT(_index, _bitmap) \
    for (_index = ucs_ffs64_safe((_bitmap)->bits[0]); \
         _index < ucs_static_array_size((_bitmap)->bits) * _UCS_BITMAP_BITS_IN_WORD; \
         _index = _UCS_BITMAP_WORD_INDEX0(_index) + \
                  ucs_ffs64_safe(_UCS_BITMAP_WORD(_bitmap, _index) & (-2ull << (uint64_t)(_index & (_UCS_BITMAP_BITS_IN_WORD - 1))))) \
        if (UCS_BITMAP_GET(_bitmap, _index))
```